third_party/move/move-vm/types/src/value_serde.rs

```diff
@@ -0,0 +1,343 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::values::{DeserializationSeed, SerializationReadyValue, SizedID, Struct, Value};
+use move_binary_format::errors::PartialVMResult;
+use move_core_types::value::{IdentifierMappingKind, MoveTypeLayout};
+use serde::{
+    de::{DeserializeSeed, Error as DeError},
+    ser::Error as SerError,
+    Deserialize, Deserializer, Serialize, Serializer,
+};
+use std::str::FromStr;
+
+pub trait CustomDeserialize {
+    fn custom_deserialize<'d, D: Deserializer<'d>>(
+        &self,
+        deserializer: D,
+        tag: &IdentifierMappingKind,
+        layout: &MoveTypeLayout,
+    ) -> Result<Value, D::Error>;
+}
+
+pub trait CustomSerialize {
+    fn custom_serialize<S: Serializer>(
+        &self,
+        serializer: S,
+        tag: &IdentifierMappingKind,
+        layout: &MoveTypeLayout,
+        sized_id: SizedID,
+    ) -> Result<S::Ok, S::Error>;
+}
+
+// FIXME: consolidate
+fn is_string_layout(layout: &MoveTypeLayout) -> bool {
```

Comment:
Should this be `is_vector_u8_layout`?

Reply:
Note that this has `if let L::Struct(move_struct) = layout` so it is not just a vector of bytes, but a layout of Move string defined as
```
struct String has copy, drop, store {
  bytes: vector<u8>,
}
```